src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -1517,7 +1534,34 @@ namespace winrt::TerminalApp::implementation
                     text = item.Path();
                 }
             }
-            eventArgs.HandleClipboardData(text);
+
+            co_await winrt::resume_foreground(Dispatcher());
+
+            const bool hasNewLine = std::find(text.cbegin(), text.cend(), L'\n') != text.cend();
+            const bool warnMultiLine = hasNewLine && _settings->GlobalSettings().WarnAboutMultiLinePaste();
+
+            constexpr const std::size_t minimumSizeForWarning = 1024 * 5; // 5 KiB
+            const bool warnLargeText = text.size() > minimumSizeForWarning &&
+                                       _settings->GlobalSettings().WarnAboutLargePaste();
+
+            if (!warnMultiLine && !warnLargeText)
+            {
+                eventArgs.HandleClipboardData(text);
+                co_return;
+            }
+
+            _acceptPaste = [=]() {
+                eventArgs.HandleClipboardData(text);
```

Comment:
Do we need to capture a weak ref to `eventArgs` here?

Reply:
No, we need a strong reference because the function returns before the user responds to the dialog so if we don't use a strong reference, the `eventArgs` will be gone.